plugins/destination/sqlite/client/migrate.go

```diff
@@ -38,70 +39,160 @@ func (i *tableInfo) getColumn(name string) *columnInfo {
-			if err := c.autoMigrateTable(ctx, table); err != nil {
-				return err
+		changes = append(changes, tableChange)
+		for _, relation := range table.Relations {
+			relationChanges, err := c.getTableChange(ctx, relation)
+			if err != nil {
+				return nil, err
 			}
-		} else {
-			c.logger.Debug().Str("table", table.Name).Msg("Table doesn't exist, creating")
-			if err := c.createTableIfNotExist(ctx, table); err != nil {
-				return err
-			}
-		}
-		if err := c.Migrate(ctx, table.Relations); err != nil {
-			return err
+			changes = append(changes, relationChanges)
 		}
 	}
-	return nil
+	return changes, nil
 }
 
-func (c *Client) isTableExistSQL(_ context.Context, table string) (bool, error) {
-	var tableExist int
-	if err := c.db.QueryRow(isTableExistSQL, table).Scan(&tableExist); err != nil {
-		return false, fmt.Errorf("failed to check if table %s exists: %w", table, err)
+func getMigrationErrors(changes []*tableChange) []string {
+	var errors []string
+	for _, tableChange := range changes {
+		if tableChange.new {
+			continue
+		}
+		for _, colChange := range tableChange.columnChanges {
+			if colChange.new && colChange.newPk {
+				errors = append(errors, fmt.Sprintf("can't migrate table %q since adding the new PK column %q is not supported. Try dropping the table", tableChange.table.Name, colChange.name))
+				// no need to report other errors as the user needs to drop the table altogether
+				continue
+			}
+			if colChange.oldType != colChange.newType {
+				errors = append(errors, fmt.Sprintf("can't migrate table %q since changing the type of column %q from %q to %q is not supported. Try dropping the column for this table", tableChange.table.Name, colChange.name, colChange.oldType, colChange.newType))
+			}
+		}
 	}
-	return tableExist == 1, nil
+	return errors
 }
 
-func (c *Client) autoMigrateTable(_ context.Context, table *schema.Table) error {
-	var err error
-	var info *tableInfo
+// This is the responsibility of the CLI of the client to lock before running migration
+func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
+	schemaChanges, err := c.getSchemaChanges(ctx, tables)
+	if err != nil {
+		return err
+	}
 
-	if info, err = c.getTableInfo(table.Name); err != nil {
-		return fmt.Errorf("failed to get table %s columns types: %w", table.Name, err)
+	migrationErrors := getMigrationErrors(schemaChanges)
+	if len(migrationErrors) > 0 {
+		return fmt.Errorf("failed to migrate schema:\n%s", strings.Join(migrationErrors, "\n"))
 	}
 
-	for _, col := range table.Columns {
-		columnName := col.Name
-		columnType := c.SchemaTypeToSqlite(col.Type)
-		sqliteColumn := info.getColumn(columnName)
-
-		switch {
-		case sqliteColumn == nil:
-			c.logger.Debug().Str("table", table.Name).Str("column", col.Name).Msg("Column doesn't exist, creating")
-			sql := "alter table \"" + table.Name + "\" add column \"" + columnName + "\" \"" + columnType + `"`
```

Comment:
We only modify columns when we add a new one and it's not a PK. I kept that logic